src/runtime/graph/graph_runtime.cc

```diff
@@ -474,23 +532,38 @@ void GraphRuntime::SetupStorage() {
     CHECK_EQ(bits % 8U, 0U);
     size_t bytes = (bits / 8U) * size;
 
-    size_t sid = static_cast<size_t>(storage_id);
-    if (sid >= pool_entry_bytes.size()) {
-      pool_entry_bytes.resize(sid + 1, 0);
-    }
-    pool_entry_bytes[sid] = std::max(pool_entry_bytes[sid], bytes);
+    uint32_t sid = static_cast<uint32_t>(storage_id);
+    device_pool_entry_bytes[sid] =
+        std::max(device_pool_entry_bytes[sid], bytes);
   }
-  // Allocate the space.
-  for (size_t i = 0; i < pool_entry_bytes.size(); ++i) {
+
+  // Allocate the space on each device.
+  for (const auto& pit : device_pool_entry_bytes) {
```

Comment:
Likely we do not need to do it per device. We can still do it by each entries in the pool, but when we try to allocate the memory in the pool, we look up which device that entry belongs to

Reply:
@tqchen But the number of iterations would be still the same, right? Going by device seems more intuitive to me.

Reply:
I agree going by device is more intuitive, but going by pool likely removes additional data structure we need( we only need a vector<NDArray> pool) and vector of the arrays. As per last comment, one of goal of runtime is to keep it as simple as possible